File: 트리/양구출작전_16437.java
package 트리;

import java.io.*;
import java.util.*;
/*
7
S 100 3
W 110 1
S 10 1
W 500 2
S 511 5
S 3 1

7
W 10 1
W 40 2
W 10 2
S 100 3
S 50 3
S 20 4
 */
public class 양구출작전_16437 {
	static int N;
	static ArrayList<Node>[] adj;
	static long[] dp;

	public static void main(String[] args) throws Exception {
	    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
	    StringTokenizer st;

		N = Integer.parseInt(br.readLine());
		adj = new ArrayList[N + 1];
		dp = new long[N + 1];

		for (int i = 0; i < N + 1; i++) {
			adj[i] = new ArrayList<>();
		}

		for (int u = 2; u <= N; u++) {
			st = new StringTokenizer(br.readLine());
			String type = st.nextToken();
			int count = Integer.parseInt(st.nextToken());
			int v = Integer.parseInt(st.nextToken());

			if (type.equals("W")) {
				adj[v].add(new Node(u, -count));
			} else {
				adj[v].add(new Node(u, count));
			}
		}

		dfs(1);
		System.out.println(dp[1]);
	}

	private static void dfs(int node) {
		for (Node child : adj[node]) {
			dp[child.here] = child.count;
			dfs(child.here);
			dp[node] += Math.max(dp[child.here], 0);
		}
	}

	private static class Node {
		int here;
		int count;

		public Node(int here, int count) {
			this.here = here;
			this.count = count;
		}
	}
}

File: 힣_코틀린_코테스터디/week8/양구출작전_16437.kt
package 힣_코틀린_코테스터디.week8

import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.*
import kotlin.math.max


class 양구출작전_16437 {
    private var N = 0
    private lateinit var adj: Array<MutableList<Node>>
    private lateinit var dp: Array<Long>

    fun solution() {
        val br = BufferedReader(InputStreamReader(System.`in`))
        var st: StringTokenizer

        N = br.readLine().toInt()
        adj = Array(N + 1) { (mutableListOf()) }
        dp = Array(N + 1) { 0L }

        for (i in 0 until N + 1) {
            adj[i] = mutableListOf()
        }

        for (u in 2..N) {
            st = StringTokenizer(br.readLine())
            val type = st.nextToken()
            val count = st.nextToken().toInt()
            val v = st.nextToken().toInt()
            if (type == "W") {
                adj[v].add(Node(u, -count))
            } else {
                adj[v].add(Node(u, count))
            }
        }
        dfs(1)
        println(dp[1])
    }

    private fun dfs(node: Int) {
        for (child in adj[node]) {
            dp[child.here] = child.count.toLong()
            dfs(child.here)
            dp[node] += max(dp[child.here], 0)
        }
    }

    private class Node(
        var here: Int,
        var count: Int
    )
}

fun main() {
    val problem = 양구출작전_16437()
    problem.solution()
}